index.js

```diff
@@ -170,87 +150,116 @@ function Jimp() {
                 parseBitmap.call(that, data, mime, cb);
             } else return throwError.call(that, "Could not load Buffer from URL <" + url + "> (HTTP: " + response.statusCode + ")", cb);
         });
-    } else if ("string" == typeof arguments[0]) {
+    } else if (typeof arguments[0] === "string") {
         // read from a path
         var path = arguments[0];
-        var cb = arguments[1];
+        cb = arguments[1];
+        options = arguments[2]; // TODO: args after cb sux.
 
-        if ("undefined" == typeof cb) cb = noop;
-        if ("function" != typeof cb)
+        if (typeof cb === "undefined") cb = noop;
+        if (typeof cb !== "function")
             return throwError.call(this, "cb must be a function", cb);
 
-        var that = this;
         getMIMEFromPath(path, function (err, mime) {
+            if (err) return throwError.call(this, err, cb);
             FS.readFile(path, function (err, data) {
                 if (err) return throwError.call(that, err, cb);
                 parseBitmap.call(that, data, mime, cb);
             });
         });
-    } else if ("object" == typeof arguments[0]) {
+    } else if (typeof arguments[0] === "object" && Buffer.isBuffer(arguments[0])) {
         // read from a buffer
         var data = arguments[0];
         var mime = getMIMEFromBuffer(data);
-        var cb = arguments[1];
+        cb = arguments[1];
+        options = arguments[2]; // TODO: args after cb sux.
 
-        if (!Buffer.isBuffer(data))
-            return throwError.call(this, "data must be a Buffer", cb);
-        if ("string" != typeof mime)
+        if (typeof mime !== "string")
             return throwError.call(this, "mime must be a string", cb);
-        if ("function" != typeof cb)
+        if (typeof cb !== "function")
             return throwError.call(this, "cb must be a function", cb);
 
-        parseBitmap.call(this, data, mime, cb);
+        parseBitmap.call(this, data, mime, cb, options);
     } else {
-        return throwError.call(this, "No matching constructor overloading was found. Please see the docs for how to call the Jimp constructor.", cb);
+        // Allow client libs to add new ways to build a Jimp object.
+        // Extra constructors must be added by `Jimp.appendConstructorOption()`
+
+        cb = arguments[arguments.length-1];
+        if (typeof cb !== "function") {
+            cb = arguments[arguments.length-2]; // TODO: try to solve the args after cb problem.
+            if (typeof cb !== "function") cb = function () {};
+        }
+        var extraConstructor = Jimp.__extraConstructors.find((c)=> c.test(...arguments));
+        if (extraConstructor)
+            new Promise(
+                (resolve, reject)=> extraConstructor.run.call(this, resolve, reject, ...arguments)
+            )
+            .then(()=> cb.call(this, null, this))
+            .catch(cb);
```

Comment:
What should probably be done instead is this:

```js
if (extraConstructor) {
    const possiblyAPromise = extraConstructor.run.call(this, ...arguments); // TODO: remove callback from arguments
    
    if (!possiblyAPromise || typeof possiblyAPromise.then !== 'function') {
        throw new TypeError('Incorrect return value of extra contructor, a Promise expected');
    }

    possiblyAPromise.then(
        ()=> setTimeout(()=> cb.call(this, null, this), 0),
        (err)=> setTimeout(()=> cb.call(this, err), 0)
    );
} else {
    /* ... */
```

This way any thrown error will remain uncaught and will crash the script and user will see his function at the top of the call stack